raiden_contracts/contracts/services/UserDeposit.sol

```diff
@@ -0,0 +1,172 @@
+    constructor(address _token_address)
+        public
+    {
+        // check token contract
+        require(_token_address != address(0x0));
+        require(contractExists(_token_address));
+        token = Token(_token_address);
+        require(token.totalSupply() > 0); // Check if the contract is indeed a token contract
+    }
+
+    /// @notice Specify trusted contracts. This has to be done outside of the
+    /// constructor to avoid cyclic dependencies.
+    /// @param _msc_address Address of the MonitoringService contract
+    function init(address _msc_address)
+        external
+    {
+        // prevent changes of trusted contract after initialization
+        require(msc_address == address(0x0));
+
+        // check monitoring service contract
+        require(_msc_address != address(0x0));
+        require(contractExists(_msc_address));
+        msc_address = _msc_address;
+    }
+
+    /// @notice Deposit tokens. Idempotent function that sets the
+    /// total_deposit of tokens of the beneficiary.
+    /// Can be called by anyone several times and on behalf of other accounts
+    /// @param beneficiary The account benefiting from the deposit
+    /// @param total_deposit The amount tokens in the user's deposit after the call
+    function deposit(address beneficiary, uint256 total_deposit)
+        external
+    {
+        require(total_deposit > balances[beneficiary]);
+
+        // Calculate the actual amount of tokens that will be transferred
+        uint256 added_deposit = total_deposit - balances[beneficiary];
+
+        balances[beneficiary] += added_deposit;
+        require(token.transferFrom(msg.sender, address(this), added_deposit));
+    }
+
+    /// @notice Internally transfer deposits between two addresses.
+    /// @param sender Account from which the amount will be deducted
+    /// @param receiver Account to which the amount will be credited
+    /// @param amount Amount of tokens to be transferred
+    /// @return true if transfer has been done successfully, otherwise false
+    function transfer(
+        address sender,
+        address receiver,
+        uint256 amount
+    )
+        canTransfer()
+        external
+        returns (bool success)
+    {
+        require(sender != receiver);
+        if (balances[sender] >= amount && amount > 0) {
+            balances[sender] -= amount;
+            balances[receiver] += amount;
+            emit BalanceReduced(sender, balances[sender]);
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    /// @notice Announce intention to withdraw tokens.
+    /// Sets the planned withdraw amount and resets the withdraw_block.
+    /// There is only one planned withdrawal at a time, the old one gets overwritten.
+    /// @param amount Maximum amount of tokens to be withdrawn
+    function planWithdraw(uint256 amount)
+        external
+    {
+        require(amount > 0);
+        require(balances[msg.sender] >= amount);
```

Comment:
I wonder if this check is necessary.  `deposit()` and `transfer()` can happen during the waiting period anyway.

Is there a document describing how UDC will be used?

Reply:
I don't think this check is strictly necessary. It might be confusing though to get events about a user planning to withdraw much more than his current balance. On the other hand, removing this limit would provide a nice way to withdraw all tokens, including potentially incoming ones, by setting a high withdraw amount.

Reply:
The UDC is meant to be used both for https://github.com/raiden-network/raiden-services/issues/38 and as a replacement of the current deposit handling inside the MSC. By unifying this, we only need a single deposit for paying both the PFS and MS.

Reply:
> removing this limit would provide a nice way to withdraw all tokens, including potentially incoming ones, by setting a high withdraw amount.

Then I would always be planning withdraws with very big numbers. What kind of game is being played here between the user and MSC? That's why I asked for a context.

Reply:
Now I see.  These withdraw plans prompt the services to claim their IOUs.  Then, actually the current implementation makes sense to me!

Reply:
The effectiveBalance is used by services to check that the user will be able to pay the service with a high likelyhood. Withdraws are immediately deducted from the effectiveBalance when they are planned, so services can claim outstanding payments during the `withdraw_delay` before the user is able to withdraw his deposit.

This is also the reason why a user might not want to plan withdraws with arbitrarily high amounts. If he does, the effectiveBalance will be zero and services will not act upon his requests.